linux-entra-sso.py

```diff
@@ -68,33 +70,35 @@ def send_message(encoded_message):
 
 
 class SsoMib:
-    NO_BROKER = {'error': 'Broker not available'}
     BROKER_NAME = 'com.microsoft.identity.broker1'
     BROKER_PATH = '/com/microsoft/identity/broker1'
     GRAPH_SCOPES = ["https://graph.microsoft.com/.default"]
 
     def __init__(self, daemon=False):
         self._bus = SessionBus()
         self.broker = None
-        self.broker_online = False
         self.session_id = uuid.uuid4()
         self._state_changed_cb = None
-        self._check_broker_online()
         if daemon:
+            self._introspect_broker(fail_on_error=False)
             self._monitor_bus()
 
-    def _check_broker_online(self):
-        dbus = self._bus.get('org.freedesktop.DBus', '/org/freedesktop/DBus')
-        if dbus.NameHasOwner(self.BROKER_NAME) \
-            or dbus.StartServiceByName(self.BROKER_NAME, 0) in \
-                [START_REPLY_ALREADY_RUNNING, START_REPLY_SUCCESS]:
-            self._instantiate_broker()
-            self.broker_online = True
-        else:
-            self.broker_online = False
-
-    def _instantiate_broker(self):
-        self.broker = self._bus.get(self.BROKER_NAME, self.BROKER_PATH)
+    def _introspect_broker(self, fail_on_error=True):
+        timeout = time.time() + BROKER_START_TIMEOUT
+        while not self.broker and time.time() < timeout:
+            try:
+                self.broker = self._bus.get(self.BROKER_NAME, self.BROKER_PATH)
+                return
+            except GLib.Error as err:
+                # GDBus.Error:org.freedesktop.dbus.errors.UnknownObject:
+                # Introspecting on non-existant object
+                # See https://github.com/siemens/linux-entra-sso/issues/33
+                if err.matches(Gio.io_error_quark(),
+                               Gio.IOErrorEnum.DBUS_ERROR):
+                    time.sleep(0.1)
```

Comment:
I think I somewhere read that there is a delay of multiple seconds until everything is up. Could that also be much short to justify this polling rate?

Reply:
I guess this depends on various conditions, but in the end we don't know. We could implement exponential backoff, but given that this is likely a bug that eventually get fixed, I don't think it is worth it. In my tests it usually took 1-3 retries. How about changing this value to 200ms instead of 100ms?

Reply:
If the broker is actually up in a few retries with 100 ms, this value is ok.